### ISSUE TITLE
Rule on interpreting JSON as LD is problematic

### ISSUE THREAD

**Issue by webb** [json-ld/json-ld.org#564]
In [section 4.9, *Interpreting JSON as JSON-LD*](https://json-ld.org/spec/latest/json-ld/#interpreting-json-as-json-ld), the rule identifying how to attach a context to a JSON instance is problematic:

> In order to use an external context with an ordinary JSON document, an author MUST specify an IRI to a valid JSON-LD document in an HTTP Link Header [RFC5988] using the http://www.w3.org/ns/json-ld#context link relation.

There are a couple of problems here:

1. The rule applies to <q>an author</q>, rather than being a constraint on an artifact, like a JSON document.

2. This rule and doesn't provide any mechanism or acceptable route for evaluating a JSON document with a separate context, other than in the context of an HTTP transfer. There should be criteria for matching the two together when they're just plain files on a disk, for example. Clarifying this might be helpful to the developer who handles the HTTP transfer correctly and ends up with a JSON document and a separate corresponding context.

**Comment by gkellogg** [comment]
@webb Perhaps you can suggest an alternative wording that would be satisfactory.

Regarding, 2), this is specifically for HTTP(S) transfers, the API provides other means of providing an expansion context which can be used for local files.

**Comment by azaroth42** [comment]
For 1. how about:

> In order to use an external context with an ordinary JSON document on the web, an IRI to a valid JSON-LD document MUST be specified in an HTTP Link Header [RFC5988] using the http:.../#context link relation when the JSON document is dereferenced.

I don't see that there's a realistic way to specify non protocol-based file-level metadata that isn't just a (hacky) file naming convention. There's no need for standardization / interoperability at that level - as @gkellogg says there's other ways to reference context documents in the processing API already.

**Comment by niklasl** [comment]
+1 for the suggestion on 1).

Regarding 2), while do I agree with @azaroth42 here, if this isn't sufficient, there is a similar mechanism with fallbacks in CSVW. See [CSVW Primer 6.7](http://w3c.github.io/csvw/primer/#well-known-location). I'm not really keen on copying it, but there might be other opinions to consider. (For one, it would make publishing JSON-LD as plain JSON with external context in static sites quite easy.)

**Comment by gkellogg** [comment]
The default metadata for CSVW was problematic, both due to naming, and the continual requirement to do a fetch to see if such a document exists, so I'd rather avoid that.

We could create protocol-specific norms, and maybe something like that for the `file` scheme would work.

**Comment by niklasl** [comment]
I agree, it would probably cause some concern to require processors to do that for every json they encounter. A norm might be nice on occasion (though it might just as well come across as a hack, as @azaroth42 says). And there is probably little benefit from standardization on that level. I'll mull a little...

**Comment by webb** [comment]
How would you feel about *completely separating* the part about HTTP headers from the concept of a JSON file being paired with a context?

From my perspective, this would make more sense to have 2 parts:

1. Define the meaning of an <q>ordinary</q> JSON text interpreted as JSON-LD with a corresponding JSON-LD context object. 

1. Establish that *one way* to pair a JSON text with a context is to link to the context in an HTTP header.

This could support, for example, having the JSON-LD playground expand a JSON blob with a given context without having to modify the JSON to accommodate the context. This would help to support the <q>Zero Edits, most of the time</q> concept outside of HTTP exchanges.
